src/array/union_partition.cc

```diff
@@ -0,0 +1,227 @@
+  for (size_t i = 0; i < csrs.size(); ++i) {
+      if (IsNullArray(csrs[i].data) == false)
+        has_data = true;
+  }
+
+  std::vector<IdArray> res_indptr;
+  std::vector<IdArray> res_indices;
+  std::vector<IdArray> res_data;
+
+  for (size_t i = 0; i < csrs.size(); ++i) {
+    aten::CSRMatrix csr = csrs[i];
+    IdArray indptr = csr.indptr + indices_offset;
+    IdArray indices = csr.indices + dst_offset;
+    if (i > 0)
+      indptr = IndexSelect(indptr,
+                           Range(1,
+                                 indptr->shape[0],
+                                 indptr->dtype.bits,
+                                 indptr->ctx));
+    res_indptr.push_back(indptr);
+    res_indices.push_back(indices);
+    src_offset += csr.num_rows;
+    dst_offset += csr.num_cols;
+
+    // any one of input csr has data index array
+    if (has_data) {
+      IdArray edges_data;
+      if (IsNullArray(csr.data)) {
+        edges_data = Range(indices_offset,
+                           indices_offset + csr.indices->shape[0],
+                           csr.indices->dtype.bits,
+                           csr.indices->ctx);
+      } else {
+        edges_data = csr.data + indices_offset;
+      }
+      res_data.push_back(edges_data);
+      indices_offset += csr.indices->shape[0];
+    }
+  }
+
+  IdArray result_indptr = Concat(res_indptr);
+  IdArray result_indices = Concat(res_indices);
+  IdArray result_data = NullArray();
+  if (has_data)
+    result_data = Concat(res_data);
+
+  return CSRMatrix(
+    src_offset, dst_offset,
+    result_indptr,
+    result_indices,
+    result_data);
+}
+
+std::vector<CSRMatrix> DisjointPartitionCsrBySizes(
+  const CSRMatrix csr,
+  const uint64_t batch_size,
+  const std::vector<uint64_t> edge_cumsum,
+  const std::vector<uint64_t> src_vertex_cumsum,
+  const std::vector<uint64_t> dst_vertex_cumsum) {
+  std::vector<CSRMatrix> ret;
+  ret.resize(batch_size);
+
+  for (size_t g = 0; g < batch_size; ++g) {
+    uint64_t num_src = src_vertex_cumsum[g+1]-src_vertex_cumsum[g];
+    IdArray result_indptr;
+    if (g == 0) {
+      result_indptr = IndexSelect(csr.indptr,
+                                  Range(0,
+                                        src_vertex_cumsum[1] + 1,
+                                        csr.indptr->dtype.bits,
+                                        csr.indptr->ctx)) - edge_cumsum[0];
+    } else {
+      IdArray result_indptr_0 = Full(0, 1,
+                                    csr.indptr->dtype.bits,
+                                    csr.indptr->ctx);
+      result_indptr = IndexSelect(csr.indptr,
+                                  Range(src_vertex_cumsum[g] + 1,
+                                        src_vertex_cumsum[g+1] + 1,
+                                        csr.indptr->dtype.bits,
+                                        csr.indptr->ctx)) - edge_cumsum[g];
+      result_indptr = Concat(std::vector<IdArray>({result_indptr_0, result_indptr}));
```

Comment:
How about slice one more value to avoid prepending a zero element?
```c++
result_indptr = IndexSelect(csr.indptr, src_vertex_cumsum[g], src_vertx_cumsum[g+1] + 1) - edge_cumsum[g];
```
This also means the two if branches could probably be merged.

Reply:
Yes. we can do that, all will - edge_cumsum[g]